### PR TITLE
feat(behavior_velocity_run_out_module): exclude obstacle crossing ego back line

### PR DESCRIPTION
**Pull request by danielsanchezaran**
## Description

With this PR an "ego back lane" is introduced, the aim of this lane is to exclude dynamic obstacles from run out consideration IF their predicted path crosses said lane. The aim is to eliminate false positive cases in which the run out module is wrongly activated when there is a pedestrian or bicycle behind the ego vehicle.

Example: the pedestrian whose predicted path touches the back lane (purple) is excluded from run out calculations. 

cap- 2024-03-25-12-00-29

## Related links

Launch changes: https://github.com/autowarefoundation/autoware_launch/pull/935

## Tests performed

PSim
Evaluator: [TIER IV INTERNAL LINK](https://evaluation.tier4.jp/evaluation/reports/ecd2efe4-1953-57b8-91ef-d78557bf344d?project_id=prd_jt)

## Notes for reviewers

Requires changes to launch: https://github.com/autowarefoundation/autoware_launch/pull/935

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [ ] The PR follows the [pull request guidelines].
- [ ] The PR has been properly tested.
- [ ] The PR has been reviewed by the code owners.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.
- [ ] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

File: planning/behavior_velocity_run_out_module/src/utils.cpp [hunk 1]
```diff
@@ -176,6 +176,61 @@ std::optional<size_t> findFirstStopPointIdx(PathPointsWithLaneId & path_points)
   return {};
 }
 
+bool pathIntersectsEgoCutLine(
+  const std::vector<geometry_msgs::msg::Pose> & path, const geometry_msgs::msg::Pose & ego_pose,
+  const double half_line_length, std::vector<geometry_msgs::msg::Point> & ego_cut_lane)
+{
+  if (path.size() < 2) return false;
+  const auto ego_rpy = rpyFromQuat(ego_pose.orientation);
```
Comment (takayuki5168) [inline planning/behavior_velocity_run_out_module/src/utils.cpp:184]: You can use tf2::getYaw(orientation).

File: planning/behavior_velocity_run_out_module/src/utils.cpp [hunk 2]
```diff
@@ -176,6 +176,61 @@ std::optional<size_t> findFirstStopPointIdx(PathPointsWithLaneId & path_points)
...
+                          const double angle) {
+    geometry_msgs::msg::Point p;
+    p.x = ego_pose.position.x + half_line_length * std::cos(angle);
+    p.y = ego_pose.position.y + half_line_length * std::sin(angle);
+    return p;
+  };
+
+  const geometry_msgs::msg::Point p1 = getOffsetPoint(ego_pose, half_line_length, ego_yaw + M_PI_2);
+  const geometry_msgs::msg::Point p2 = getOffsetPoint(ego_pose, half_line_length, ego_yaw - M_PI_2);
+  ego_cut_lane = {p1, p2};
```
Comment (takayuki5168) [inline planning/behavior_velocity_run_out_module/src/utils.cpp:198]: You can use tier4_autoware_utils::calcOffsetPose(ref_pose, x, y, z). Then, you don't have to calculate yaw here.

File: planning/behavior_velocity_run_out_module/src/utils.cpp [hunk 3]
```diff
@@ -176,6 +176,61 @@ std::optional<size_t> findFirstStopPointIdx(PathPointsWithLaneId & path_points)
...
+//   const auto ego_rpy = rpyFromQuat(ego_pose.orientation);
+//   const double ego_yaw = ego_rpy.z;
+//   const Point2d p1{
+//     ego_pose.position.x + offset * std::cos(ego_yaw + M_PI_2),
+//     ego_pose.position.y + offset * std::sin(ego_yaw + M_PI_2)};
+//   const Point2d p2{
+//     ego_pose.position.x + offset * std::cos(ego_yaw - M_PI_2),
+//     ego_pose.position.y + offset * std::sin(ego_yaw - M_PI_2)};
+//   return {p1, p2};
+// }
```
Comment (takayuki5168) [inline planning/behavior_velocity_run_out_module/src/utils.cpp:232]: can be removed?

File: planning/behavior_velocity_run_out_module/src/utils.cpp [hunk 4]
```diff
@@ -181,20 +181,15 @@ bool pathIntersectsEgoCutLine(
...
-    p.y = ego_pose.position.y + half_line_length * std::sin(angle);
-    return p;
-  };
-
-  const geometry_msgs::msg::Point p1 = getOffsetPoint(ego_pose, half_line_length, ego_yaw + M_PI_2);
-  const geometry_msgs::msg::Point p2 = getOffsetPoint(ego_pose, half_line_length, ego_yaw - M_PI_2);
+  auto getOffsetPoint =
+    [](const geometry_msgs::msg::Pose & ego_pose, const double half_line_length) {
+      const auto p = tier4_autoware_utils::calcOffsetPose(ego_pose, 0.0, half_line_length, 0.0);
+      return tier4_autoware_utils::createPoint(p.position.x, p.position.y, p.position.z);
```
Comment (takayuki5168) [inline planning/behavior_velocity_run_out_module/src/utils.cpp:187]:
```cpp
tier4_autoware_utils::calcOffsetPose(ego_pose, 0.0, half_line_length, 0.0).position
```
is enough.

Comment (danielsanchezaran) [reply planning/behavior_velocity_run_out_module/src/utils.cpp]: @takayuki5168 Ah! right thanks for the review

File: planning/behavior_velocity_run_out_module/src/utils.hpp [hunk 5]
```diff
@@ -194,6 +204,27 @@ struct DynamicObstacleData
...
+  return rpy;
+}
+
+bool pathIntersectsEgoCutLine(
+  const std::vector<geometry_msgs::msg::Pose> & path, const geometry_msgs::msg::Pose & ego_pose,
+  const double half_line_length, std::vector<geometry_msgs::msg::Point> & ego_cut_lane);
+
+// LineString2d createLineStringFromPath(const PredictedPath & path);
+
+// LineString2d createEgoCutLine(const geometry_msgs::msg::Pose & ego_pose, const double offset);
```
Comment (TomohitoAndo) [inline planning/behavior_velocity_run_out_module/src/utils.hpp:226]: These comments can be removed?

File: planning/behavior_velocity_run_out_module/src/debug.hpp [hunk 6]
```diff
@@ -134,12 +135,14 @@ class RunOutDebug
...
   std::vector<geometry_msgs::msg::Point> collision_points_;
   std::vector<geometry_msgs::msg::Point> nearest_collision_point_;
+  std::vector<geometry_msgs::msg::Point> ego_back_lane_;
   std::vector<geometry_msgs::msg::Pose> stop_pose_;
   std::vector<std::vector<geometry_msgs::msg::Point>> predicted_vehicle_polygons_;
   std::vector<std::vector<geometry_msgs::msg::Point>> predicted_obstacle_polygons_;
   std::vector<std::vector<geometry_msgs::msg::Point>> collision_obstacle_polygons_;
   std::vector<std::vector<geometry_msgs::msg::Point>> detection_area_polygons_;
   std::vector<std::vector<geometry_msgs::msg::Point>> mandatory_detection_area_polygons_;
+
```
Comment (TomohitoAndo) [inline planning/behavior_velocity_run_out_module/src/debug.hpp:145]: Could you please remove the unnecessary blank lines?

Comment (danielsanchezaran) [reply planning/behavior_velocity_run_out_module/src/debug.hpp]: Done @TomohitoAndo Thanks!

File: planning/behavior_velocity_run_out_module/README.md [hunk 7]
```diff
@@ -108,6 +108,14 @@ You can choose whether to use this feature by parameter of `use_partition_lanele
 
 ![brief](./docs/exclude_obstacles_by_partition.svg)
 
+##### Exclude obstacles that cross the ego vehicle's "cut line"
+
+This module can exclude can exclude obstacles that have predicted paths that come from behind the ego vehicle. It excludes obstacles if the their predicted path crosses the ego's "cut line". The "cut line" is generating a virtual line segment that is perpendicular to the ego vehicle and that passes through the ego's base link.
```
Comment (kosuke55) [inline planning/behavior_velocity_run_out_module/README.md:113]:
```suggestion
This module can exclude obstacles that have predicted paths that come from behind the ego vehicle. It excludes obstacles if their predicted path crosses the ego's "cut line". The "cut line" is generating a virtual line segment that is perpendicular to the ego vehicle and that passes through the ego's base link.
```

File: planning/behavior_velocity_run_out_module/src/scene.cpp [hunk 8]
```diff
@@ -102,9 +102,18 @@ bool RunOutModule::modifyPathVelocity(
...
+    // extract obstacles using lanelet information
+    const auto partition_excluded_obstacles =
+      excludeObstaclesOutSideOfPartition(dynamic_obstacles, extended_smoothed_path, current_pose);
+
+    // extract obstacles that cross the ego's back lane
+    if (!planner_param_.run_out.use_ego_cut_line) return partition_excluded_obstacles;
+
+    const auto ego_back_line_excluded_obstacles =
+      excludeObstaclesCrossingEgoBackLine(partition_excluded_obstacles, current_pose);
+    return ego_back_line_excluded_obstacles;
```
Comment (kosuke55) [inline planning/behavior_velocity_run_out_module/src/scene.cpp:115]:
```suggestion
    if (!planner_param_.run_out.use_ego_cut_line) return partition_excluded_obstacles;

    // extract obstacles that cross the ego's back lane
    const auto ego_back_line_excluded_obstacles =
      excludeObstaclesCrossingEgoBackLine(partition_excluded_obstacles, current_pose);
    return ego_back_line_excluded_obstacles;
```

Comment (kosuke55) [reply planning/behavior_velocity_run_out_module/src/scene.cpp:115]: the comment place is better like this?

File: planning/behavior_velocity_run_out_module/README.md (hunk 7 above)
Comment (kosuke55) [reply planning/behavior_velocity_run_out_module/README.md:113]: > This module can exclude obstacles that have predicted paths that come from BEHIND the ego vehicle.

the exclued targets are all objects that cross the cut line, not just the BEHIND?

Comment (danielsanchezaran) [reply planning/behavior_velocity_run_out_module/README.md]: You are right, the explanation was not clear, it excludes obstacles if they pass the back part of the ego by crossing the line. I think it is better now, please confirm @kosuke55

File: planning/behavior_velocity_run_out_module/src/scene.cpp (hunk 8 above)
Comment (danielsanchezaran) [reply planning/behavior_velocity_run_out_module/src/scene.cpp]: Done, thanks for the suggestion! @kosuke55

**Review by kosuke55 — APPROVED** [review]
LGTM! thanks!!

**Pull request merged**